File: src/calculator/paypal_calculator.rs
use crate::{
    calculator::{ChargeAmount, SaleBreakdown},
    store::config,
};

use super::Material;

const PAYMENT_PROCESSING_PERCENTAGE: f64 = 0.015;
const EEA_PAYMENT_PROCESSING_PERCENTAGE: f64 = 0.025;
const INTERNATIONAL_PAYMENT_PROCESSING_PERCENTAGE: f64 = 0.035;
const PAYMENT_PROCESSING_FEE: f64 = 0.3;

pub(crate) fn based_on_sale(
    sale: f64,
    delivery_costs: f64,
    eea: bool,
    international: bool,
) -> SaleBreakdown {
    let config = config::get_config();

    let sale_total = sale + delivery_costs;
    let additional_payment_processing_percentage = if eea {
        EEA_PAYMENT_PROCESSING_PERCENTAGE
    } else if international {
        INTERNATIONAL_PAYMENT_PROCESSING_PERCENTAGE
    } else {
        0.0
    };
    let base_processing_payment =
        (sale_total * PAYMENT_PROCESSING_PERCENTAGE) + PAYMENT_PROCESSING_FEE;
    let additional_processing_payment =
        base_processing_payment * additional_payment_processing_percentage;
    let payment_processing_cost = base_processing_payment + additional_processing_payment;
    let tax = sale * (config.tax_rate / 100.0);
    let revenue = sale - payment_processing_cost - tax;
    let percentage_kept = (revenue / sale) * 100.0;
    let max_working_hours = revenue / config.hourly_rate;
    SaleBreakdown {
        sale,
        delivery_costs,
        transaction_cost: 0.0,
        payment_processing_cost,
        offsite_ads_cost: 0.0,
        regulatory_operating_fee: 0.0,
        vat_paid_by_buyer: 0.0,
        vat_on_seller_fees: 0.0,
        total_fees: payment_processing_cost,
        total_fees_with_vat: payment_processing_cost,
        tax,
        revenue,
        percentage_kept,
        max_working_hours,
    }
}

pub(crate) fn how_much_to_charge(
    number_of_minutes: f64,
    material_costs: Vec<Material>,
    delivery_costs: f64,
    eea: bool,
    international: bool,
) -> ChargeAmount {
    let config = config::get_config();

    let total_material_costs: f64 = material_costs.iter().map(|material| material.value).sum();
    let base_charge =
        ((number_of_minutes / 60.0) * config.hourly_rate) + total_material_costs + delivery_costs;
    let additional_payment_processing_percentage = if eea {
        EEA_PAYMENT_PROCESSING_PERCENTAGE
    } else if international {
        INTERNATIONAL_PAYMENT_PROCESSING_PERCENTAGE
    } else {
        0.0
    };
    let base_processing_payment =
        (base_charge * PAYMENT_PROCESSING_PERCENTAGE) + PAYMENT_PROCESSING_FEE;
    let additional_processing_payment =
        base_processing_payment * additional_payment_processing_percentage;
    let payment_processing_cost = base_processing_payment + additional_processing_payment;
    let charge = base_charge + payment_processing_cost;

    let markup_percentage = (100.0 + config.markup_percentage) / 100.0;

    let total_to_charge = (charge * markup_percentage).ceil();
    ChargeAmount {
        total_to_charge,
        with_vat: total_to_charge * 1.2,
    }
}

File: composeApp/src/commonMain/kotlin/com/benjaminsproule/costcalculator/calculator/PaypalCalculator.kt
package com.benjaminsproule.costcalculator.calculator

import com.benjaminsproule.costcalculator.store.Config
import kotlin.math.ceil

enum class PaypalLocation(override val label: String) : Option {
    Local("Local"),
    EEA("EEA"),
    International("International"),
}

data class PaypalSale(
    override val cost: Float,
    override val deliveryCosts: Float,
    val location: PaypalLocation
) : Sale()

data class PaypalCharge(
    override val numberOfMinutes: Float,
    override val materialCosts: List<Material>,
    override val deliveryCosts: Float,
    val location: PaypalLocation
) : Charge()

class PaypalCalculator(private val config: Config) : Calculator<PaypalSale, PaypalCharge> {
    companion object {
        const val PAYMENT_PROCESSING_PERCENTAGE: Float = 0.015f
        const val EEA_PAYMENT_PROCESSING_PERCENTAGE: Float = 0.025f
        const val INTERNATIONAL_PAYMENT_PROCESSING_PERCENTAGE: Float = 0.035f
        const val PAYMENT_PROCESSING_FEE: Float = 0.3f
    }

    override fun basedOnSale(sale: PaypalSale): SaleBreakdown {
        val saleTotal = sale.cost + sale.deliveryCosts
        val additionalPaymentProcessingPercentage = if (sale.location == PaypalLocation.EEA) {
            EEA_PAYMENT_PROCESSING_PERCENTAGE
        } else if (sale.location == PaypalLocation.International) {
            INTERNATIONAL_PAYMENT_PROCESSING_PERCENTAGE
        } else {
            0.0f
        }
        val baseProcessingPayment = (saleTotal * PAYMENT_PROCESSING_PERCENTAGE) + PAYMENT_PROCESSING_FEE
        val additionalProcessingPayment = saleTotal * additionalPaymentProcessingPercentage
        val paymentProcessingCost = baseProcessingPayment + additionalProcessingPayment;
        val tax = sale.cost * (config.taxRate / 100.0f)
        val revenue = sale.cost - paymentProcessingCost - tax
        val percentageKept = (revenue / sale.cost) * 100.0f
        val maxWorkingHours = revenue / config.hourlyRate
        return SaleBreakdown(
            sale = sale.cost,
            deliveryCosts = sale.deliveryCosts,
            transactionCost = 0.0f,
            paymentProcessingCost = paymentProcessingCost,
            offsiteAdsCost = 0.0f,
            regulatoryOperatingFee = 0.0f,
            vatPaidByBuyer = 0.0f,
            vatOnSellerFees = 0.0f,
            totalFees = paymentProcessingCost,
            totalFeesWithVat = paymentProcessingCost,
            tax = tax,
            revenue = revenue,
            percentageKept = percentageKept,
            maxWorkingHours = maxWorkingHours,
        )
    }

    override fun howMuchToCharge(
        charge: PaypalCharge
    ): ChargeAmount {
        val totalMaterialCosts = charge.materialCosts.map { material -> material.value }.sum();
        val baseCharge =
            ((charge.numberOfMinutes / 60.0f) * config.hourlyRate) + totalMaterialCosts + charge.deliveryCosts;
        val additionalPaymentProcessingPercentage = if (charge.location == PaypalLocation.EEA) {
            EEA_PAYMENT_PROCESSING_PERCENTAGE
        } else if (charge.location == PaypalLocation.International) {
            INTERNATIONAL_PAYMENT_PROCESSING_PERCENTAGE
        } else {
            0.0f
        }
        val baseProcessingPayment =
            (baseCharge * PAYMENT_PROCESSING_PERCENTAGE) + PAYMENT_PROCESSING_FEE;
        val additionalProcessingPayment =
            baseCharge * additionalPaymentProcessingPercentage;
        val paymentProcessingCost = baseProcessingPayment + additionalProcessingPayment;
        val chargeWithFees = baseCharge + paymentProcessingCost;

        val markupPercentage = (100.0f + config.markupPercentage) / 100.0f

        val totalToCharge = ceil(chargeWithFees * markupPercentage)
        return ChargeAmount(
            totalToCharge = totalToCharge,
            withVat = totalToCharge * 1.2f,
        )
    }
}
